File: frontend/src/api/bootstrap.ts
/**
 * Typed data layer for the first-run installation wizard (session 6.4).
 * These routes are public and unauthenticated — no CSRF header needed on reads,
 * but the apiClient adds it to POST calls automatically via the double-submit cookie.
 * We use rawRequest (no auth retry) because there are no auth cookies yet.
 */

export interface BootstrapStatus {
  needs_setup: boolean
}

export interface PreflightCheck {
  name: string
  ok: boolean
  detail: string
  hint?: string | null
}

export interface PreflightResult {
  checks: PreflightCheck[]
  all_ok: boolean
}

export interface AdminIn {
  email: string
  full_name: string
  password: string
}

export interface BrandingIn {
  product_name?: string
  default_tz?: string
  accent_hex?: string | null
}

export interface NotificationsIn {
  smtp_host?: string | null
  smtp_port?: number
  smtp_from?: string | null
  smtp_username?: string | null
  smtp_password?: string | null
}

export interface CompleteIn {
  admin: AdminIn
  branding?: BrandingIn
  notifications?: NotificationsIn | null
}

export interface CompleteOut {
  setup_complete: boolean
  user_id: number
  email: string
}

async function _post<T>(path: string, body?: unknown): Promise<T> {
  const headers: Record<string, string> = { 'Content-Type': 'application/json' }
  const response = await fetch(path, {
    method: 'POST',
    headers,
    credentials: 'same-origin',
    body: body === undefined ? undefined : JSON.stringify(body),
  })
  if (!response.ok) {
    const text = await response.text().catch(() => response.statusText)
    throw { status: response.status, detail: text }
  }
  if (response.status === 204) return undefined as T
  return response.json() as Promise<T>
}

export const bootstrapApi = {
  status: () => fetch('/api/bootstrap/status').then((r) => r.json() as Promise<BootstrapStatus>),

  preflight: () => _post<PreflightResult>('/api/bootstrap/preflight'),

  complete: (payload: CompleteIn) => _post<CompleteOut>('/api/bootstrap/complete', payload),
}

File: backend/app/schemas/bootstrap.py
"""Pydantic schemas for the bootstrap / first-run wizard API (session 6.4)."""

from pydantic import BaseModel, EmailStr, Field, field_validator


class BootstrapStatusOut(BaseModel):
    needs_setup: bool


class PreflightCheck(BaseModel):
    name: str
    ok: bool
    detail: str
    hint: str | None = None


class PreflightOut(BaseModel):
    checks: list[PreflightCheck]
    all_ok: bool


class AdminIn(BaseModel):
    email: EmailStr
    full_name: str = Field(min_length=1, max_length=120)
    password: str = Field(min_length=8, max_length=128)


class BrandingIn(BaseModel):
    product_name: str = Field(default="FDM Platform", min_length=1, max_length=80)
    default_tz: str = Field(default="Asia/Dubai", max_length=64)
    accent_hex: str | None = Field(default=None, max_length=7)

    @field_validator("accent_hex")
    @classmethod
    def validate_hex(cls, v: str | None) -> str | None:
        if v is None:
            return v
        if not v.startswith("#") or len(v) not in (4, 7):
            raise ValueError("accent_hex must be #RGB or #RRGGBB")
        return v


class NotificationsIn(BaseModel):
    smtp_host: str | None = None
    smtp_port: int = 587
    smtp_from: str | None = None
    smtp_username: str | None = None
    smtp_password: str | None = None


class CompleteIn(BaseModel):
    admin: AdminIn
    branding: BrandingIn = Field(default_factory=BrandingIn)
    notifications: NotificationsIn | None = None


class CompleteOut(BaseModel):
    setup_complete: bool
    user_id: int
    email: str
